includes/class-wc-order-item-coupon.php

```diff
@@ -0,0 +1,187 @@
+			$offset = 'discount';
+		} elseif ( 'discount_amount_tax' === $offset ) {
+			$offset = 'discount_tax';
+		}
+		return parent::offsetGet( $offset );
+	}
+
+	/**
+	 * offsetSet for ArrayAccess/Backwards compatibility.
+	 * @deprecated Add deprecation notices in future release.
+	 * @param string $offset
+	 * @param mixed $value
+	 */
+	public function offsetSet( $offset, $value ) {
+		if ( 'discount_amount' === $offset ) {
+			$offset = 'discount';
+		} elseif ( 'discount_amount_tax' === $offset ) {
+			$offset = 'discount_tax';
+		}
+		parent::offsetSet( $offset, $value );
+	}
+
+	/**
+	 * offsetExists for ArrayAccess
+	 * @param string $offset
+	 * @return bool
+	 */
+	public function offsetExists( $offset ) {
+		if ( in_array( $offset, array( 'discount_amount', 'discount_amount_tax' ) ) ) {
+			return true;
+		}
+		return parent::offsetExists( $offset );
+	}
+
+	/**
+	 * Read/populate data properties specific to this order item.
+	 */
+	public function read( $id ) {
+		parent::read( $id );
+		if ( $this->get_id() ) {
+			$this->set_discount( get_metadata( 'order_item', $this->get_id(), 'discount_amount', true ) );
+			$this->set_discount_tax( get_metadata( 'order_item', $this->get_id(), 'discount_amount_tax', true ) );
+		}
+	}
+
+	/**
+	 * Save properties specific to this order item.
+	 * @return int Item ID
+	 */
+	public function save() {
+		parent::save();
+		if ( $this->get_id() ) {
+			wc_update_order_item_meta( $this->get_id(), 'discount_amount', $this->get_discount() );
+			wc_update_order_item_meta( $this->get_id(), 'discount_amount_tax', $this->get_discount_tax() );
+		}
+
+		return $this->get_id();
+	}
+
+	/**
+	 * Internal meta keys we don't want exposed as part of meta_data.
+	 * @return array()
+	 */
+	protected function get_internal_meta_keys() {
+		return array( 'discount_amount', 'discount_amount_tax' );
+	}
+
+	/*
+	|--------------------------------------------------------------------------
+	| Setters
+	|--------------------------------------------------------------------------
+	*/
+
+	/**
+	 * Set order item name.
+	 * @param string $value
+	 */
+	public function set_name( $value ) {
+		$this->set_code( $value );
+	}
+
+	/**
+	 * Set code.
+	 * @param string $value
+	 */
+	public function set_code( $value ) {
+		$this->_data['code'] = wc_clean( $value );
+	}
+
+	/**
+	 * Set discount amount.
+	 * @param string $value
+	 */
+	public function set_discount( $value ) {
+		$this->_data['discount'] =  wc_format_decimal( $value );
+	}
+
+	/**
+	 * Set discounted tax amount.
+	 * @param string $value
+	 */
+	public function set_discount_tax( $value ) {
+		$this->_data['discount_tax'] = wc_format_decimal( $value );
+	}
+
+	/*
+	|--------------------------------------------------------------------------
+	| Getters
+	|--------------------------------------------------------------------------
+	*/
+
+	/**
+	 * Get order item type.
+	 * @return string
+	 */
+	public function get_type() {
+		return 'coupon';
```

Comment:
Hardcoding the return value of `get_type()` here and in the other the `WC_Order_Item_*` classes breaks the `WC_Order_Item->set_type()` API which allows the type property and return value of `get_type()` to be changed.

I see why it's hardcoded - the type is linked to the class. But it's definitely a code smell and will create gotchas, like this:

```
$line_item = WC_Order_Factory::get_order_item( $coupon_line_item_id );
echo $line_item->get_type();             // will print 'coupon'
$line_item->set_type( 'shipping' );    // won’t throw a warning or error
echo $line_item->get_type();             // still prints 'coupon' even those it was changed
```

I can't really see a clean solution. Because honouring the type property with `get_type()` could also lead to confusion, e.g.

```
$line_item = WC_Order_Factory::get_order_item( $coupon_line_item_id );
echo $line_item->get_type();             // will print 'coupon'
$line_item->set_type( 'product' );
echo $line_item->get_type();             // now prints 'product'

if ( 'product' == $line_item->get_type() ) {
	// Cool, looks like we have a product line item, so let's call a method only available on WC_Order_Item_Product
	$download = $line_item->get_item_downloads(); // Uh oh WSOD, should have checked $line_item instanceof WC_Order_Item_Product instead
}
```